code/modules/pai/software.dm

```diff
@@ -0,0 +1,396 @@
+			if(aicamera)
+				aicamera.toggle_camera_mode(usr)
+				return TRUE
+			return FALSE
+		if("printer module")
+			if(aicamera)
+				aicamera.paiprint(usr)
+				return TRUE
+			return FALSE
+		if("radio")
+			if(radio)
+				radio.attack_self(src)
+				return TRUE
+			return FALSE
+		if("refresh")
+			if(params["list"] == "security" && !installed_software.Find("security records") \
+				|| params["list"] == "medical" && !installed_software.Find("medical records"))
+				return FALSE
+			refresh_records(ui, params["list"])
+			return TRUE
+		if("remote signaler")
+			if(!signaler)
+				signaler = new(src)
+			signaler.ui_interact(src)
+			return TRUE
+		if("security HUD")
+			toggle_hud(usr, "security")
+			return TRUE
+		if("universal translator")
+			grant_languages(usr, ui)
+			return TRUE
+	return FALSE
+
+/**
+ * Purchases the selected software from the list and deducts their
+ * available ram.
+ *
+ * @param {mob} user The user purchasing the software.
+ * @param {string} selection The software to purchase.
+ * @return {bool} TRUE if the software was purchased, FALSE otherwise.
+ */
+/mob/living/silicon/pai/proc/buy_software(mob/user, selection)
+	if(!available_software.Find(selection) || installed_software.Find(selection))
+		to_chat(user, span_warning("Error: Software unavailable."))
+		CRASH("[user] tried to purchase unavailable software as a pAI.")
+	var/cost = available_software[selection]
+	if(ram < cost)
+		to_chat(user, span_warning("Error: Insufficient RAM available."))
+		CRASH("[user] tried to purchase software with insufficient RAM.")
+	installed_software.Add(selection)
+	ram -= cost
+	var/datum/hud/pai/pAIhud = hud_used
+	pAIhud?.update_software_buttons()
+	return TRUE
+
+/**
+ * Changes the image displayed on the pAI.
+ *
+ * @param {mob} user The user who is changing the image.
+ * @return {bool} TRUE if the image was changed, FALSE otherwise.
+ */
+/mob/living/silicon/pai/proc/change_image(mob/user)
+	var/new_image = tgui_input_list(user, "Select your new display image", \
+		"Display Image", possible_overlays)
+	if(isnull(new_image))
+		return FALSE
+	card.emotion_icon = new_image
+	card.update_appearance()
+	return TRUE
+
+/**
+ * Supporting proc for the pAI to prick it's master's hand
+ * or... whatever. It must be held in order to work
+ * Gives the owner a popup if they want to get the jab.
+ *
+ * @param {mob} user The pAI requesting the sample.
+ * @param {living/carbon} master The holder of the pAI.
+ * @return {boolean} TRUE if a sample was taken, FALSE otherwise.
+ */
+/mob/living/silicon/pai/proc/check_dna(mob/user)
+	var/mob/living/carbon/holder = get_holder()
+	if(!holder)
+		to_chat(user, span_warning("You must be in someone's hands to do this!"))
+		return FALSE
+	to_chat(user, span_notice("Requesting a DNA sample."))
+	if(tgui_alert(holder, "[user] is requesting a DNA sample from you. \
+		Will you allow it to confirm your identity?", "Checking DNA", \
+		list("Yes", "No")) != "Yes")
```

Comment:
the constant use of backslash newlines really makes it harder to read I find, I'm against your overuse of it.

Reply:
needs a standard of some sort. i will remove all of my usages.